### PR TITLE
Fix GHASH-ASM implementation on s390x

### PR DESCRIPTION
**Pull request by juergenchrist**
s390x GHASH assembler implementation assumed it was called from a
gcm128_context structure where the Xi paramter to the ghash function was
embedded in that structure.  Since the structure layout resembles the paramter
block required for kimd-GHASH, the assembler code simply assumed the 128 bytes
after Xi are the hash subkey.

This assumption was broken with the introduction of AES-GCM-SIV which uses the
GHASH implementation without a gcm128_context structure.  Furthermore, the
bytes following the Xi input parameter to the GHASH function do not contain
the hash subkey.  To fix this, we remove the assumption about the calling
context and build the parameter block on the stack.  This requires some
copying of data to and from the stack.  While this introduces a performance
degradation, new systems anyway use kma for GHASH/AES-GCM.

Finally fixes #18693 for s390x.

Signed-off-by: Juergen Christ <jchrist@linux.ibm.com>

### CODE REVIEW

**Comment by tmshort** [comment]
Thanks for this. If you want to go with this (even with the performance change), I can review it.
Alternatively, I can put in the missing fields into the AES-GCM-SIV structure (even if it means pulling in the larger `GCM128_CONTEXT` structure), which ought to avoid modifying the s390x assembly.

**Comment by juergenchrist** [comment]
Personally I do not like the assumption in the s390x assembler code.  Predates my time with s390x.  But I just assume code like that written in C (e.g. using `offsetof` to get a pointer to `gcm128_context` from the function parameter `Xi`) would not be allowed in.

Also, the performance impact is minimal.  The last two generations of s390x machines do not use the ghash function for AES-GCM or even GHASH.  So I guess that the performance impact is very limited, but the code is cleaner.  Thus I prefer the version from this PR.

**Comment by tmshort** [comment]
> Also, the performance impact is minimal.

OK, I'll keep my update in my back pocket in case some *other* assembly has problems with AES-GCM-SIV.

**Review by tmshort — APPROVED** [review]
Based on my *just-learned* s390x assembly syntax, this loads the value of `Xi` (argument) and a part of the `Htable` (argument, which presumably would've been in the `H` field of the `GCM128_CONTEXT`, and copies them into the stack. The address from the stack is used instead of `Xi`. Since the output is put onto the stack, it needs to be copied back into `Xi`.
This is a 32-byte initial copy (four `uint64_t`'s/registers), followed by a 16-byte copy of the results back into `Xi`.

**Comment by tmshort** [comment]
> Personally I do not like the assumption in the s390x assembler code. Predates my time with s390x. But I just assume code like that written in C (e.g. using `offsetof` to get a pointer to `gcm128_context` from the function parameter `Xi`) would not be allowed in.

So, it appears what's happening in the s390x assembly is that it is assuming that `Xi` is really a structure to `{ u64 Xi[2]; u64 H[2]; }`, and not really using `offsetof`. I don't really consider it *that* bad. Now, if it had been doing anything more (e.g. using a negative offset or a greater positive offset), then I would be in 100% agreement with you.

**Comment by juergenchrist** [comment]
In s390x, kimd-GHASH requires a structure of the form
```
struct kimdghashparam {
  uint64_t chainingvalue[2]; // Xi used as input and output
  uint64_t hashsubkey[2];
};
```
If you look at `gcm128_context`, these fields are already there:
```
struct gcm128_context {
    /* Following 6 names follow names in GCM specification */
    union {
        u64 u[2];
        u32 d[4];
        u8 c[16];
        size_t t[16 / sizeof(size_t)];
    } Yi, EKi, EK0, len, Xi, H;
    /*
     * Relative position of Yi, EKi, EK0, len, Xi, H and pre-computed Htable is
     * used in some assembler modules, i.e. don't change the order!
     */
    u128 Htable[16];
    struct gcm_funcs_st funcs;
    unsigned int mres, ares;
    block128_f block;
    void *key;
#if !defined(OPENSSL_SMALL_FOOTPRINT)
    unsigned char Xn[48];
#endif
};
```
So the old assembler code just loaded the address of the passed `Xi` paramter assuming it is part of a `gcm128_context` structure that contains `H` directly after `Xi`.  This structure is then used as parameter block to kimd-GHASH.  There, `Xi` is used as chaining value, `H` as hash subkey, and the input is processed until there is no more block left.  The resulting hash is placed back into `Xi`.

In the new assembler code, I create this parameter block on the stack by copying `Xi` and the hash subkey which is located in `Htable[8]` and the use the stack address as parameter block location.  With that, the hash ends up on the stack.  So I have to copy it back to the output location.

**Comment by tmshort** [comment]
@juergenchrist yup, that's what I got out of reading the updated code. I had misread the original assembly comments and conflated `H` and `Htable`, which is why my other fix did not work.

**Comment by juergenchrist** [comment]
> which is why my other fix did not work

I just added a comment to #18911 which would fix your implementation of polyval via ghash and use ghash in the style of `gcm128_context`.  For some reason you used `Xi` as input value and not as chaining value.  With the positions of `Xi` and `out` switched, we could keep the old assembler code.  But that already shows how fragile and complicated to use this construct currently is.

**Comment by tmshort** [comment]
I pushed the (corrected) alternative AES-GCM-SIV-code-only fix, mostly for posterity.

**Comment by tmshort** [comment]
@juergenchrist look as though you removed some commented out code?

**Comment by juergenchrist** [comment]
> look as though you removed some "if 0'd" out code?

Yes, I had the choice to either fix this, too, or simply remove it.  I do not see a point in "fixing" unused code.  So I decided to remove it.

**Comment by tmshort** [comment]
ping @openssl/otc for review?

File: crypto/modes/asm/ghash-s390x.pl
Comment (mattcaswell) [inline crypto/modes/asm/ghash-s390x.pl:127]: Generally we prefer to remove dead code. I'm not sure what benefit leaving this here adds.

Comment (juergenchrist) [reply crypto/modes/asm/ghash-s390x.pl:127]: Yes, I will remove this.  Left in it for easier review and explanation of the change.

Comment (juergenchrist) [reply crypto/modes/asm/ghash-s390x.pl]: Removed the code and moved the comment to the correct position.

**Comment by tmshort** [comment]
Reconfirm

**Comment by juergenchrist** [comment]
What is the status here?  I still see AES-GCM-SIV failing in our internal CI on a real s390x.

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
> What is the status here? I still see AES-GCM-SIV failing in our internal CI on a real s390x.

It was waiting on my approval - now done. Now we just need to wait for the 24hr timer to expire before it is eligible for merge.

**Comment by mattcaswell** [comment]
Pushed to master. Thanks.

**Pull request closed without merging**
